pkg/controller/datavolume-controller.go

```diff
@@ -1236,6 +1243,80 @@ func (r *DatavolumeReconciler) newPersistentVolumeClaim(dataVolume *cdiv1.DataVo
 				}),
 			},
 		},
-		Spec: *dataVolume.Spec.PVC,
+		Spec: *pvcSpec,
 	}, nil
 }
+
+func (r *DatavolumeReconciler) renderPvcSpec(dv *cdiv1.DataVolume) (*corev1.PersistentVolumeClaimSpec, error) {
+	// right now Spec.PVC is required and tested before, so no need to check if its null here!
+	pvcSpecCopy := dv.Spec.PVC.DeepCopy()
+	storageClass, err := GetStorageClassByName(r.client, dv.Spec.PVC.StorageClassName)
+	if err != nil {
+		return nil, err
+	}
+	if storageClass == nil {
+		// Not even default storageClass on the cluster, cannot apply the defaults
+		return pvcSpecCopy, nil
+	}
+
+	if len(pvcSpecCopy.AccessModes) == 0 {
+		accessMode, err := getDefaultAccessMode(r.client, storageClass)
+		if err != nil {
+			r.log.V(1).Info("Cannot set accessMode for new pvc", "namespace", dv.Namespace, "name", dv.Name)
+			r.recorder.Eventf(dv, corev1.EventTypeWarning, ErrClaimNotValid,
+				fmt.Sprintf("DataVolume spec is missing accessMode and cannot get access mode from StorageProfile %s", storageClass.Name))
+			return nil, err
+		}
+		pvcSpecCopy.AccessModes = append(pvcSpecCopy.AccessModes, *accessMode)
+	}
+	if pvcSpecCopy.VolumeMode == nil || *pvcSpecCopy.VolumeMode == "" {
+		volumeMode, err := getDefaultVolumeMode(r.client, storageClass)
+		if err != nil {
+			return nil, err
+		}
+		pvcSpecCopy.VolumeMode = volumeMode
+	}
+
+	return pvcSpecCopy, nil
+}
+
+func getDefaultVolumeMode(c client.Client, storageClass *storagev1.StorageClass) (*corev1.PersistentVolumeMode, error) {
+	storageProfile, err := getStorageProfile(c, storageClass.Name)
+	if err != nil {
+		// TODO: should it default to nil and just ignore the error for optional params like volumeMode when not found?
+		// and only return some errors with communication
```

Comment:
Hmm, this is a system error not a user error so I think we can still raise it.  Failure to fetch a Profile could be because we are creating it.  In that case we should be trying again during the next reconciliation loop.

Reply:
that's right, done